### DISCUSSION TITLE
FBX file format: go/no go

### DISCUSSION THREAD

**Discussion by lonelyhawk** [Markemp/Cryengine-Converter#111] [General]
what are the plans regarding FBX file format? It's proprietary and not well-documented. The text-oriented one can be implemented.  but the binary one is tricky. Looks overcomplicated. (I can build a marshaling lib, see https://github.com/lonelyhawk/ManagedFBX )

**Comment by Markemp** [comment]
It's something I've wanted to do for a while.  3d file formats are really not in a good place.  Collada stopped work at v1.5, and Blender only supports up to v1.4.1.  And Blender doesn't really do much to support Collada either; it works, but nobody is bug fixing or enhancing it.  FBX is proprietary and documentation is questionable.  Alembic doesn't support animations, and I'm not even sure it's being developed anymore either.

My big pie in the sky dream state would be to learn c++ good enough to take over the Collada support in Blender, but that's a heavy lift.

I *want* FBX support, but not sure if anybody *needs* it right now.  Are there any specific use cases for implementing FBX in the near future?  Long term, will FBX support be required if Collada support disappears for 3DS Max and Blender?

**Comment by vmxeo** [comment]
Ugh, FBX.

- Alembic would be my first choice, as it's still widely adopted. It does support animation (so long as it's baked).

- FBX would be the next choice, as it's also just as widely adopted, but as stated it's also proprietary (Autodesk at that)

- Collada is still my choice, but Blender doesn't directly support it's own importer. This is a nice breakdown of what's wrong: https://code.blender.org/2016/10/the-collada-case/

- You could write it directly out to a Blender file, though not sure how users of other render software would like that.

**Comment by Kjasi** [comment]
Maybe we should add the USD file format? It seems to hit a lot of the requirements, is open source, and it's gaining in popularity and support. Most 3D programs already support it, including Blender. https://graphics.pixar.com/usd/docs/index.html

**Reply by Markemp** [reply]
I don't think Blender has a USD importer; just an exporter?  https://docs.blender.org/manual/en/latest/files/import_export/usd.html

**Reply by vmxeo** [reply]
The USD importer just popped out of the oven a few weeks ago: https://developer.blender.org/tag/usd/

Not if it's in the 3.0 alpha, or the nightlies...

**Reply by Markemp** [reply]
Looks like they just started supporting development for it.  And since they have a full time dev (apparently NVidia is providing them) to provide support, I think that it may be the preferred method going forward.

Currently their API only supports c++ and python.  I think I'm going to make a USD.Net project to work like the Grendgine plug-in I use for Collada to simplify porting things over.  But yeah, going to keep an eye on this format, since it ticks a lot of the boxes I'm looking for.

**Reply by Markemp** [reply]
Does USD support animations or particle systems?  I know they aren't implemented yet in the converter, but it would be nice if the animation .caf files could be included in a .usd file.

It seems like it can also make scenes.  I wonder how the hierarchy works.  Can you have a scene file that references a bunch of assets, and then automatically build from that?

**Reply by Kjasi** [reply]
Yes, the USD format supports Animation. Particles too (I think) but only "active" particles. At least, that's what the Blender 2.93 Exporter page is saying. It also says they don't support spot lights, but the format is open, so we could add that if we need it.

**Reply by lonelyhawk** [reply]
spent the month trying to write the .blend exporter. if you haven't seen the most perverted file format - read the docs on .blend. should have gone with fbx instead, for sure.

**Reply by Markemp** [reply]
Hehe, I spent a day or two at it a while back.  It's not a very portable format.  .blend files just do too much stuff.
